### DISCUSSION TITLE
Dualize: Error using |, operands must be real

### DISCUSSION THREAD

**Discussion by alastair-abbott** [yalmip/YALMIP#1169] [Possible bug or missing functionality] [answered]
Hi,

Yalmip has been crashing on some relatively large SDP problems I have when dualize is set to 1 (for smaller problems of this type I have found faster convergence using scs-direct with this option, which is why I'm using it here).

I get the error:

```
Error using  | 
Operands must be real.

Error in dualize>detect_diagonal_terms (line 875)
    candidates = find((B(:,1) == 0) & (sum(B | B,2) == 1) & (sum(B,2) == 1));

Error in dualize (line 175)
implicit_positive = detect_diagonal_terms(F);
```

On closer inspection, we can see the issue is that B has an epsilon imaginary part:

```
K>> imag(B)

ans =

   1.0e-17 *

         (1,1)             0.3578
       (128,1)             0.3578
```

Can we generically remove the imaginary part of B, or use a more robust way of converting it to a logical here?

**Comment by johanlofberg** [comment]
Can you post a reproducible example

**Reply by alastair-abbott** [reply]
I've attached a self-contained example that is reproducible for me (i.e., running dualize_issue_example.mat). 

But unfortunately I don't have a simpler example; even within the class of SDPs I have, it only seems to arise in rather large examples.

I'm running MATLAB R2021b if ever it's relevant
[dualize_error.zip](https://github.com/yalmip/YALMIP/files/9358735/dualize_error.zip).

**Comment by johanlofberg** [comment, marked as answer]
There should be no imaginary part there, so the fix will be an addition of B = real(B) on line 874 (unless I clean it upstream)
